File: leetcode/2976. Minimum Cost to Convert String I/main.go
package main

import "math"

const (
	EngChars = 26
)

func minimumCost(source string, target string, original []byte, changed []byte, cost []int) int64 {
	graph := buildGraph(original, changed, cost)
	optimizePaths(graph)
	return computeTotalCost(source, target, graph)
}

func buildGraph(original, changed []byte, cost []int) [][]int {
	graph := make([][]int, EngChars)
	for i := range graph {
		graph[i] = make([]int, EngChars)
		for j := range graph[i] {
      graph[i][j] = math.MaxInt
		}
	}

  for i, c := range cost {
    from := int(original[i] - 'a')
    to := int(changed[i] - 'a')

    if c < graph[from][to] {
      graph[from][to] = c
    }
  }

  return graph
}

func optimizePaths(graph [][]int) {
  for k := 0; k < EngChars; k++ {
    for i := 0; i < EngChars; i++ {
      if graph[i][k] < math.MaxInt {
        for j := 0; j < EngChars; j++ {
          if graph[k][j] < math.MaxInt {
            if graph[i][k] + graph[k][j] < graph[i][j] {
              graph[i][j] = graph[i][k] + graph[k][j]
            }
          }
        }
      }
    }
  }
}

func computeTotalCost(source, target string, graph [][]int) int64 {
  var total int64

  for i := 0; i < len(source); i++ {
    sourceChar := int(source[i] - 'a')
    targetChar := int(target[i] - 'a')

    if sourceChar != targetChar {
      if graph[sourceChar][targetChar] == math.MaxInt {
        return -1
      }

      total += int64(graph[sourceChar][targetChar])
    }
  }

  return total
}

File: leetcode/2976. Minimum Cost to Convert String I/main.rs
use std::cmp::min;
use std::i64;

impl Solution {
    pub fn minimum_cost(
        source: String,
        target: String,
        original: Vec<char>,
        changed: Vec<char>,
        cost: Vec<i32>,
    ) -> i64 {
        let mut graph = Self::build_graph(&original, &changed, &cost);
        Self::optimize_graph(&mut graph);
        Self::compute_total_cost(&source, &target, &graph)
    }

    fn build_graph(original: &[char], changed: &[char], cost: &[i32]) -> Vec<Vec<i64>> {
        let mut graph = vec![vec![i64::MAX; 26]; 26];
        for i in 0..26 {
            graph[i][i] = 0;
        }
        for (i, &o) in original.iter().enumerate() {
            let from = (o as usize) - ('a' as usize);
            let to = (changed[i] as usize) - ('a' as usize);
            graph[from][to] = min(graph[from][to], cost[i] as i64);
        }

        graph
    }

    fn optimize_graph(graph: &mut Vec<Vec<i64>>) {
        for k in 0..26 {
            for i in 0..26 {
                if graph[i][k] < i64::MAX {
                    for j in 0..26 {
                        if graph[k][j] < i64::MAX {
                            if graph[k][j] < i64::MAX {
                                graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j]);
                            }
                        }
                    }
                }
            }
        }
    }

    fn compute_total_cost(source: &str, target: &str, graph: &[Vec<i64>]) -> i64 {
        if source.len() != target.len() {
            return -1;
        }

        let mut total = 0;

        for (s, t) in source.chars().zip(target.chars()) {
            let source_char = (s as usize) - ('a' as usize);
            let target_char = (t as usize) - ('a' as usize);

            if source_char != target_char {
                if graph[source_char][target_char] == i64::MAX {
                    return -1;
                }

                total += graph[source_char][target_char];
            }
        }

        total
    }
}

